tests/SchedulerTest.php

```diff
@@ -0,0 +1,221 @@
+    }
+
+    protected function runOff()
+    {
+        Loop::futureTick(function () {
+            Loop::stop();
+        });
+
+        Loop::run();
+    }
+
+    public function testSchedulingTasks()
+    {
+        $task = new PromiseBoundTask(Promise\resolve());
+        $now = new DateTime();
+        $now->modify('+2 month');
+
+        $nextDue = null;
+        $this->scheduler
+            ->on(CountableScheduler::ON_TASK_SCHEDULED, function (Task $_, DateTime $time) use (&$nextDue) {
+                $nextDue = $time;
+            })
+            ->schedule($task, new AbsoluteDueFrequency($now));
+
+        $this->runOff();
+
+        $this->assertEquals($now, $nextDue, 'Scheduler could not get the correct schedule date time');
+    }
+
+    public function testNeverDueTaskSchedules()
+    {
+        $neverRun = true;
+        $this->scheduler
+            ->schedule(new PromiseBoundTask(Promise\resolve()), new NeverDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $_, ExtendedPromiseInterface $p) use (&$neverRun) {
+                $neverRun = false;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($neverRun, 'Scheduler::schedule() runs a task with a never due frequency');
+    }
+
+    public function testRunningScheduledTasks()
+    {
+        $hasRun = false;
+        $task = new PromiseBoundTask(Promise\resolve());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $t, ExtendedPromiseInterface $_) use (&$hasRun) {
+                $hasRun = true;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($hasRun, 'Scheduler::schedule() could not run scheduled tasks');
+    }
+
+    public function testCancelingRunningTasks()
+    {
+        $promisesCount = [];
+        $task = new PromiseBoundTask((new Promise\Deferred())->promise());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_CANCEL, function (Task $_, array $promises) use (&$promisesCount) {
+                $promisesCount = $promises;
+            });
+
+        // We have to use addtimer!!
+        Loop::addTimer(0, function () use ($task) {
+            $this->scheduler->remove($task);
+        });
+
+        $this->runOff();
+
+        $this->assertCount(2, $promisesCount, 'Scheduler::remove() could not cancel running tasks');
+    }
+
+    public function testFailedTasks()
+    {
+        $taskFailed = false;
+        $deferred = new Promise\Deferred();
+        $task = new PromiseBoundTask($deferred->promise());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $_, ExtendedPromiseInterface $p) use ($deferred) {
+                $deferred->reject(new TaskRejectedException('Exited mysteriously!'));
```

Comment:
I would also test that this is the exception you are getting in the callback. Validate the instance and the message please. I would only use "rejected" for the latter.